main/driver_setup.c

```diff
@@ -0,0 +1,409 @@
+#include "driver_setup.h"
+
+#include "driver/gpio.h"
+#include "driver/twai.h"
+#include "esp_check.h"
+#include "esp_eth.h"
+#include "esp_eth_driver.h"
+#include "esp_event.h"
+#include "esp_log.h"
+#include "esp_mac.h"
+#include "esp_netif.h"
+#include "esp_wifi.h"
+#include "memory.h"
+
+// A pointer to the ethernet netif object.
+// NULL if ethernet wasn't started.
+esp_netif_t *driver_setup_eth_netif = NULL;
+
+// A pointer to the WIFI netif object.
+// NULL if WIFI wasn't started.
+esp_netif_t *driver_setup_wifi_netif = NULL;
+
+// Static memory for the can_recovery_task stack.
+static StackType_t can_recovery_task_stack[4096];
+
+// Static memory for the can_recovery_task.
+static StaticTask_t can_recovery_task_mem;
+
+// Semaphore that specifies when the internet is ready
+static SemaphoreHandle_t internet_ready = NULL;
+
+// Name that will be used for logging
+static const char *TAG = "driver_setup";
+
+// Event handler purely for logging debugging information.
+static void ethernet_event_handler(void *arg, esp_event_base_t event_base,
+                                   int32_t event_id, void *event_data);
+static void wifi_event_handler(void *arg, esp_event_base_t event_base,
+                               int32_t event_id, void *event_data);
+static void ip_event_handler(void *arg, esp_event_base_t event_base,
+                             int32_t event_id, void *event_data);
+
+// A FreeRTOS task that gets spawned by
+// `driver_setup_can()`.
+static void can_recovery_task(void *pvParameters);
+
+// Prints the status of `netif` to `buf` in JSON format.
+// Returns the number of bytes written, or -1 if ran out of space.
+static int print_netif_status(esp_netif_t *netif, char *buf_out, size_t buflen);
+
+void driver_setup_ethernet(const esp_netif_ip_info_t *ip_info) {
+  // Based on:
+  // https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/network/esp_eth.html
+
+  // Check if the ethernet driver has already been started
+  if (driver_setup_eth_netif != NULL) {
+    ESP_LOGE(TAG, "Can only call driver_setup_ethernet() one time. Aborting.");
+    abort();
+  }
+
+  //// Create an ethernet MAC object ////
+  eth_mac_config_t mac_config = ETH_MAC_DEFAULT_CONFIG();
+  eth_esp32_emac_config_t esp32_emac_config = ETH_ESP32_EMAC_DEFAULT_CONFIG();
+  // 23 nased on ESP32-EVB schematic
+  esp32_emac_config.smi_mdc_gpio_num = 23;
+  // 18 based on ESP32-EVB schematic
+  esp32_emac_config.smi_mdio_gpio_num = 18;
+  esp_eth_mac_t *mac = esp_eth_mac_new_esp32(&esp32_emac_config, &mac_config);
+  assert(mac);
```

Comment:
Asserts are generally good for things where you expect that they should logically never happen. I kind of almost think of them as hints to the compiler that a path won't be taken instead of error handling. That's because if you compile with `#define NDEBUG`, suddenly your asserts will disappear.

Since it looks like this does a memory alloc behind the scenes, I'd recommend a different error handling pattern like an explicit call to a panic function that won't be accidentally disabled.

Reply:
Good point. I fixed this by making the driver initialization functions recoverable in 041181b.